File: allowanced-fs/src/FileProvider.fs
module FileProviders

open Scriban
open System.IO
open System.Collections.Immutable
open Microsoft.Extensions.FileProviders

type IFileProvider =
    abstract member getFileContent: string -> string
    abstract member getTemplate: string -> Scriban.Template

type LocalFileProvider(root: string) =
    interface IFileProvider with
        member this.getFileContent(path: string): string = 
            File.ReadAllText(Path.Combine(root, path))
        
        member this.getTemplate(path: string): Template = 
            let content = (this :> IFileProvider).getFileContent(path)
            Template.Parse(content)

type EmbeddedResourceProvider(assembly: System.Reflection.Assembly, root: string) =
    let embeddedFileProvider = new EmbeddedFileProvider(assembly, root)
    let mutable templateCache = ImmutableDictionary.Create<string, Template>()

    interface IFileProvider with
        member this.getFileContent(path: string): string = 
            let fileInfo = embeddedFileProvider.GetFileInfo(path)
            use stream = fileInfo.CreateReadStream()
            use reader = new StreamReader(stream)
            reader.ReadToEnd()
        
        member this.getTemplate(path: string): Template = 
            match templateCache.TryGetValue(path) with
            | true, template -> template
            | false, _ ->
                let content = (this :> IFileProvider).getFileContent(path)
                let template = Template.Parse(content)
                templateCache <- templateCache.Add(path, template)
                template

File: allowanced/FileProvider.cs
using System.Collections.Immutable;
using System.IO;
using System.Security.Cryptography.X509Certificates;
using Microsoft.Extensions.FileProviders;
using Scriban;

namespace allowanced
{
    public interface IFileProvider
    {
        string getFileContents(string fileName);
        Scriban.Template getTemplate(string fileName);
    }

    public record LocalFileProvider(string root) : IFileProvider
    {
        public string getFileContents(string fileName)
        {
            return File.ReadAllText(Path.Combine(root, fileName));
        }

        public Template getTemplate(string fileName)
        {
            var content = getFileContents(fileName);
            return Template.Parse(content);
        }
    }

    public record EmbeddedResourceProvider(System.Reflection.Assembly assembly, string root) : IFileProvider
    {
        EmbeddedFileProvider embeddedFileProvider = new EmbeddedFileProvider(assembly, root);
        ImmutableDictionary<string, Template> templateCache = ImmutableDictionary<string, Template>.Empty;

        public string getFileContents(string fileName)
        {
            var fileInfo = embeddedFileProvider.GetFileInfo(fileName);
            using var stream = fileInfo.CreateReadStream();
            using var reader = new StreamReader(stream);
            return reader.ReadToEnd();
        }

        public Template getTemplate(string fileName)
        {
            if (templateCache.TryGetValue(fileName, out var template))
            {
                return template;
            }
            else
            {
                var content = getFileContents(fileName);
                template = Template.Parse(content);
                templateCache = templateCache.Add(fileName, template);
                return template;
            }
        }
    }

    public static class FileProviderExtensions {
        public static IServiceCollection ConfigureFileProvider(this IServiceCollection services) {
            #if DEBUG
                services.AddSingleton<IFileProvider>(new LocalFileProvider("www"));
            #else
                services.AddSingleton<IFileProvider>(new embeddedFileProvider(
                    System.Reflection.Assembly.GetExecutingAssembly(),
                    "allowanced/www")
            #endif
            return services;
        }
    }
}
